bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java

```diff
@@ -550,11 +594,452 @@ private static RetryOptions createRetryOptions(Configuration configuration) {
+  }
+
+  // copied over from CredentialFactory,
+  // TODO: Need to find better way to convert P12 key into Credentials
+  private static Credentials getCredentialFromPrivateKeyServiceAccount(
+      String serviceAccountEmail, String privateKeyFile) throws IOException {
+    try {
+      PrivateKey privateKey =
+          SecurityUtils.loadPrivateKeyFromKeyStore(
+              SecurityUtils.getPkcs12KeyStore(),
+              new FileInputStream(privateKeyFile),
+              "notasecret",
+              "privatekey",
+              "notasecret");
+
+      return ServiceAccountJwtAccessCredentials.newBuilder()
+          .setClientEmail(serviceAccountEmail)
+          .setPrivateKey(privateKey)
+          .build();
+    } catch (GeneralSecurityException exception) {
+      throw new RuntimeException("exception while retrieving credentials", exception);
+    }
+  }
+
+  /** Creates {@link TransportChannelProvider} for plaintext negotiation type. */
+  private static TransportChannelProvider buildPlainTextChannelProvider(
+      String endpoint, int channelCount) {
+    return defaultGrpcTransportProviderBuilder()
+        .setEndpoint(endpoint)
+        .setPoolSize(channelCount)
+        .setChannelConfigurator(
+            new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+              @Override
+              public ManagedChannelBuilder apply(ManagedChannelBuilder channelBuilder) {
+                return channelBuilder.usePlaintext();
+              }
+            })
+        .build();
+  }
+
+  /** Creates {@link Set} of {@link StatusCode.Code} from {@link Status.Code} */
+  private static Set<StatusCode.Code> buildRetryCodes(Configuration configuration) {
+    ImmutableSet.Builder<StatusCode.Code> statusCodeBuilder = ImmutableSet.builder();
+
+    for (Status.Code retryCode : RetryOptions.DEFAULT_ENABLE_GRPC_RETRIES_SET) {
+      statusCodeBuilder.add(GrpcStatusCode.of(retryCode).getCode());
+    }
+
+    String retryCodes = configuration.get(ADDITIONAL_RETRY_CODES, "");
+    String[] codes = retryCodes.split(",");
+    for (String stringCode : codes) {
+      String trimmed = stringCode.trim();
+      if (trimmed.isEmpty()) {
+        continue;
+      }
+      StatusCode.Code code = StatusCode.Code.valueOf(trimmed);
+      Preconditions.checkNotNull(code, "Code " + stringCode + " not found.");
+      statusCodeBuilder.add(code);
+      LOG.debug("gRPC retry on: %s", stringCode);
+    }
+
+    return statusCodeBuilder.build();
+  }
+
+  /** Creates {@link RetrySettings} for non-streaming idempotent method. */
+  private static RetrySettings buildIdempotentRetrySettings(
+      RetrySettings originalRetrySettings, Configuration configuration) {
+
+    RetrySettings.Builder retryBuilder = originalRetrySettings.toBuilder();
+    if (configuration.getBoolean(ALLOW_NO_TIMESTAMP_RETRIES_KEY, false)) {
+      throw new UnsupportedOperationException("Retries without Timestamp does not support yet.");
+    }
+
```

Comment:
Retries without Timestamp does not support yet.
=>
No Timestamp retries are not supported yet.

Where do you check, that timestamp is not defined?

Reply:
> Where do you check, that timestamp is not defined?

The timestamp logic is in the bigtable-core module([here](https://github.com/googleapis/java-bigtable-hbase/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java#L224-L235)). But with veneer client this logic is pushed back to gax, and we do not have any control to enable this check.

Reply:
I'm asking because you throw an exception if ALLOW_NO_TIMESTAMP_RETRIES_KEY == false, but you do not check if TIMESTAMP == null. 
If you think this is ok, I'm ok )

Reply:
The timestamp would be picked while a user creates `Mutation` object.

The core-client allows mutations without a timestamp. With veneer client, each row mutation must have a timestamp(either client-side or server-side) attached to it.

Reply:
Oh, yes. I see. `buildIdempotentRetrySettings()` checks not the retry, but the settings. And it will fail if one tries to create settings which allow no timestamp retries. 
Thanks for explanation.

I have no other questions